### ISSUE TITLE
Clarification on the Evaluation Methodology of the Baseline Method(s)

### ISSUE THREAD

**Issue by huahuaai** [KANABOON1/MemGen#15]
thank you for open-sourcing this great work and for sharing the code and results. I really appreciate the effort you've put into this project!

But I'm still confused about some things. I am unable to reproduce the same results as those in the paper using the dataset processing method described in the code.  what‘s the vanilla, Cot
and SFT's setting, which is important for me to compare your method with other methods.

**Comment by huahuaai** [comment]
In your data processor the prompt of gsm8k is CoT setting，what is the vanilla  and sft 's config.

**Comment by KANABOON1** [comment]
https://github.com/KANABOON1/MemGen/blob/e1880dbd3d9b0213f0b0f77122e853ff38945d0a/data/gsm8k/builder.py#L43

Hi! I rechecked our current implementation. The prompt template we use is indeed the vanilla format — we do not apply any `think step by step` or similar CoT-style prompting by default.

For SFT, we keep exactly the same vanilla prompt template as in the released code. For the CoT setting, we explicitly enable step-by-step reasoning using prompts like `think step by step`. As mentioned in our README roadmap, the baseline CoT and SFT implementation will be released later in a separate repository.

**Comment by KANABOON1** [comment]
@huahuaai 
I just re-tested GSM8K with the same vanilla prompt on another cluster, and the accuracy was 40.79%, which is very close to the results reported in our paper.
I just use: 0-shot, temperature = 0 and max_response_length = 4096.

Could you share your exact settings? Or did you obtain your results through an API?

**Comment by huahuaai** [comment]
> [@huahuaai](https://github.com/huahuaai) I just re-tested GSM8K with the same vanilla prompt on another cluster, and the accuracy was 40.79%, which is very close to the results reported in our paper. I just use: 0-shot, temperature = 0 and max_response_length = 4096.
> 
> Could you share your exact settings? Or did you obtain your results through an API?

`import os
import argparse
import torch
import random
import numpy as np
import logging
import jsonlines
from tqdm import tqdm
import json
import time
import os
import pdb
from torch.utils.data import DataLoader
from datasets import Dataset
from accelerate import Accelerator
from datetime import datetime
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer
)
from transformers import PreTrainedTokenizerBase
from trl import SFTTrainer, SFTConfig, GRPOConfig
from trl.models import unwrap_model_for_generation

from common.config import Config
from common.logger import setup_logger
from data import get_data_builder
from data.utils.math_utils import compute_score


def set_seed(random_seed: int, use_gpu: bool):

    random.seed(random_seed)
    os.environ['PYTHONHASHSEED'] = str(random_seed)
    np.random.seed(random_seed)
    torch.manual_seed(random_seed)
    torch.cuda.manual_seed(random_seed)
    if use_gpu:
        torch.cuda.manual_seed_all(random_seed)

    torch.backends.cudnn.deterministic = True   
    torch.backends.cudnn.benchmark = False      

    print(f"set seed: {random_seed}")

def parse_args():
    parser = argparse.ArgumentParser(description="Language Reasoning and Memory")

    parser.add_argument("--cfg-path", required=True, help="path to configuration file.")
    parser.add_argument(
        "--options",
        nargs="+",
        help="override some settings in the used config, the key-value pair "
        "in xxx=yyy format will be merged into config file (deprecate), "
        "change to --cfg-options instead.",
    )

    args = parser.parse_args()

    return args

def main():
        # parse configs
    args = parse_args()
    config = Config(args)
    
    set_seed(config.run_cfg.seed, use_gpu=True)
    # build components
    config_dict = config.to_dict()
    data_builder = get_data_builder(config_dict.get("dataset"))
    dataset_dict = data_builder.get_dataset_dict()
    print(dataset_dict)
    env_cls = data_builder.get_env_cls()
    env = env_cls(config_dict.get("dataset"))

    
    test_dataset = dataset_dict["test"]

    model = AutoModelForCausalLM.from_pretrained(
        "./Qwen2.5-1.5B-Instruct", 
        torch_dtype=torch.bfloat16, 
        attn_implementation="flash_attention_2"
    )
    tokenizer = AutoTokenizer.from_pretrained("./Qwen2.5-1.5B-Instruct")

    accelerator = Accelerator()
    batch_size = 32
    output_dir = "results/single"

    # prepare dataset and dataloader
    test_dataloader = accelerator.prepare(DataLoader(
        dataset=test_dataset, 
        batch_size=8, 
        shuffle=False,
        collate_fn=lambda batch: batch
    ))

    # prepare model
    model_wrapped = accelerator.prepare_model(model=model, evaluation_mode=True)
    model_wrapped.eval()
    scores = []
    # batch generate
    for step, test_batch in enumerate(tqdm(test_dataloader)):
        with unwrap_model_for_generation(
            model_wrapped, accelerator
        ) as unwrapped_model:
            prompts = [x["prompt"] for x in test_batch]
            labels = [x["solution"] for x in test_batch]
            messages = [[
                {
                    "role": "user",
                    "content": prompt
                }
            ] for prompt in prompts]
            inputs = [tokenizer.apply_chat_template(message, tokenize=False, add_generation_prompt=True) for message in messages]
            model_inputs = tokenizer(inputs, return_tensors="pt", padding=True, padding_side="left", add_special_tokens=True).to(accelerator.device)
            generated_ids = model.generate(
                **model_inputs,
                max_new_tokens=1024,
            )
            generated_ids = [
                output_ids[len(input_ids):] for input_ids, output_ids in zip(model_inputs.input_ids, generated_ids)
            ]

            responses = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
            for response, ground_truth in zip(responses, labels):
                score = compute_score(response, ground_truth)
                scores.append(score)
                with jsonlines.open(os.path.join(output_dir, 'single_wo_sft_new.jsonl'), mode='a') as writer:
                    writer.write({
                        'response': response,
                        'ground_truth': ground_truth,
                        'score': score
                    })
    result = {"acc": sum(scores)/len(scores)}
    print(result)
    with open(os.path.join(output_dir, 'single_wo_sft_new.txt'), 'w') as f:
        json.dump(result, f)


if __name__ == "__main__":
    main()`  
this is my evaluate script just use your metric function and the same dataset procesor. and my max_new_tokens is 1024

**Comment by huahuaai** [comment]
> [@huahuaai](https://github.com/huahuaai) I think I may have figured out the reason. It’s because you applied tokenizer.apply_chat_template before passing the input to the model. For single-turn datasets, we do not use any chat template. This is likely why your results are higher than ours. You can refer to our MemGenRunner implementation: for multi-turn datasets we enable the chat template, but for single-turn datasets we do not enable it (for both the baseline and MemGen, in order to keep the comparison fair).
> 
> [MemGen/memgen/runner.py](https://github.com/KANABOON1/MemGen/blob/e1880dbd3d9b0213f0b0f77122e853ff38945d0a/memgen/runner.py#L34)
> 
> Line 34 in [e1880db](/KANABOON1/MemGen/commit/e1880dbd3d9b0213f0b0f77122e853ff38945d0a)
> 
>  class MemGenRunner:

Thank you for your patient reply. I'll adjust my configuration and try again.

**Comment by zelinzeng** [comment]
> [@zelinzeng](https://github.com/zelinzeng)
> 
> 1. At the moment, the MemGen code cannot serve as a vanilla baseline, because it will automatically initialize a Weaver model if load_model_path is set to None. So by default, you are not using a pure vanilla model.
> 2. (Haha, thanks for sharing the results — it’s interesting that you obtained a higher score than what we reported in our paper. This shows MemGen works quite well, right?) One possible reason is that you used the ChatML template, which can generally lead to better performance compared to our reported results. In our experiments, neither the baseline nor MemGen used ChatML, so the comparison is fair. However, I do have a concern: the base Weaver model is untrained, so injecting an untrained Weaver should actually degrade the Reasoner’s performance. It should not outperform the SFT-trained Weaver. This makes me suspect there might be an issue with how the model weights were loaded on your side. Could this be related to your loading procedure?

@KANABOON1 
Thanks! To be sure that I loaded the weights correctly, I modified the loading logic. Here is my implementation:

### Base Weaver + Trigger (GRPO)

`evaluate_logs`
<img width="953" height="461" alt="Image" src="https://github.com/user-attachments/assets/bd3c8247-ad09-4631-9744-7ea7831aaaf3" />

`answer.json`
<img width="868" height="258" alt="Image" src="https://github.com/user-attachments/assets/0352744c-9302-4b5b-a72d-a3194043f54a" />

`trainer_logs`
<img width="746" height="509" alt="Image" src="https://github.com/user-attachments/assets/2e1adcb2-b2bf-44e6-85cf-8c667faa1e37" />

### Weaver SFT + Trigger (GRPO)
`evaluate_logs`
<img width="939" height="506" alt="Image" src="https://github.com/user-attachments/assets/249111f4-bbb0-4147-90c9-72c359865260" />

`answer.json`
<img width="881" height="255" alt="Image" src="https://github.com/user-attachments/assets/bb73898b-a5f9-4058-9a61-b7b5a3025c8e" />

`train_trigger_logs`
<img width="936" height="457" alt="Image" src="https://github.com/user-attachments/assets/90e2ef30-ef99-43ea-a262-b158d660f172" />

Could you please take a brief look at the screenshot above? Does this align with how you intended the weights to be composed during inference? 

Thanks again for your time!

**Comment by KANABOON1** [comment]
I couldn’t see the issue from your screenshot. Could I take a look at your trigger training logs?
